### PR TITLE
build(Examples,Tools): Upgrade RISC-V Toolchain, Update Project Files

### PR DESCRIPTION
**Pull request by Jake-Carter**
This PR depends on #619 for its consolidated gcc.mk files.

## Overview

- Add support for xPack `riscv-none-elf` toolchain.  (https://github.com/Analog-Devices-MSDK/msdk/commit/5188f32e6b9e7b281487a0ed7e3defcc229f1710)
- Use `riscv-none-elf` toolchain as default instead of the deprecated `riscv-none-embed`.  (https://github.com/Analog-Devices-MSDK/msdk/commit/d678542640ea3628de1f7d675eae6de7665d51a5)
- Update project files to use `riscv-none-elf` version [12.2.0-3.1](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/releases/tag/v12.2.0-3.1).  (https://github.com/Analog-Devices-MSDK/msdk/commit/d678542640ea3628de1f7d675eae6de7665d51a5, https://github.com/Analog-Devices-MSDK/msdk/pull/628/commits/92d8d1c7fdacee5050481fefc6fd8af95dd6f7e9)
  - The install location for the updated toolchain shall be `Tools/xPack/riscv-none-elf-gcc/12.2.0-3.1`
- Add the new toolchain to the build tester
- Update all project files

### Project File Updates

For the most part, the changes are:
- Use the new RISC-V toolchain
```diff
diff --git a/Examples/MAX78000/Hello_World/.vscode/settings.json b/Examples/MAX78000/Hello_World/.vscode/settings.json
index 3a6b363122..8e185bb9ce 100755
--- a/Examples/MAX78000/Hello_World/.vscode/settings.json
+++ b/Examples/MAX78000/Hello_World/.vscode/settings.json
@@ -28,11 +28,11 @@
     "RV_OCD_target_file":"${config:target}_riscv.cfg",
 
     "v_Arm_GCC":"10.3",
-    "v_xPack_GCC":"10.2.0-1.2",
+    "v_xPack_GCC":"12.2.0-3.1",
 
     "OCD_path":"${config:MAXIM_PATH}/Tools/OpenOCD",
     "ARM_GCC_path":"${config:MAXIM_PATH}/Tools/GNUTools/${config:v_Arm_GCC}",
-    "xPack_GCC_path":"${config:MAXIM_PATH}/Tools/xPack/riscv-none-embed-gcc/${config:v_xPack_GCC}",
+    "xPack_GCC_path":"${config:MAXIM_PATH}/Tools/xPack/riscv-none-elf-gcc/${config:v_xPack_GCC}",
     "Make_path":"${config:MAXIM_PATH}/Tools/MSYS2/usr/bin",
```
- Remove duplicate entries in `.vscode/settings.json` (resolves [MSDK-1201](https://jira.maxim-ic.com/browse/MSDK-1201))
- Include a file for detecting compiler-specific and build-specific compiler definitions (see [comment](https://github.com/Analog-Devices-MSDK/msdk/pull/613#issuecomment-1591866010)) (https://github.com/Analog-Devices-MSDK/msdk/pull/628/commits/53527bbcf6a1d622a5e15c83ccef2b4bc9693f50, https://github.com/Analog-Devices-MSDK/msdk/pull/628/commits/35ab3dfe6d833e28b4bdb0868d83e1ec116215db)
```diff
diff --git a/Examples/MAX32520/AES/.vscode/settings.json b/Examples/MAX32520/AES/.vscode/settings.json
index d0310e8209..df1171e0f7 100755
--- a/Examples/MAX32520/AES/.vscode/settings.json
+++ b/Examples/MAX32520/AES/.vscode/settings.json
@@ -71,6 +71,9 @@
         "TARGET=${config:target}",
         "TARGET_REV=0x4131",
         "__GNUC__"
+    ],
+    "C_Cpp.default.forcedInclude": [
+        "${workspaceFolder}/build/project_defines.h"
     ]
 }
```

MSDKGen has also been updated to check for projects that come preconfigured for non-default boards to match project settings when needed.

image

## RISC-V -march Change

Commit:  https://github.com/Analog-Devices-MSDK/msdk/pull/628/commits/5188f32e6b9e7b281487a0ed7e3defcc229f1710

The upgrade to `riscv-none-elf` changed the default ISA spec.  As recommended by [this](https://groups.google.com/a/groups.riscv.org/g/sw-dev/c/aE1ZeHHCYf4) discussion post, the default `-march` option when using the new toolchain is now  `-march=rv32im_zicsr_zifencei`.  Note the new option does not use the compressed instruction set.

The old toolchain still defaults to `-march=rv32imc`.

`-mabi` has been kept at `ilp32`.

Our old `gcc_riscv.mk` file also seemed to be passing `-march=rv32im` to the _compiler_, but `-march=rv32imafdc` to the _linker_.  It also did not give the `-mabi` option to the linker.  I think this was a mistake but I'm not sure what the ramifications of the old behavior was.  The new implementation passes the same `-march` and `-mabi` to the compiler, assembler, and linker stages.

### CODE REVIEW

**Review by sihyung-maxim — APPROVED** [review]
Don't know too much about the toolchain and project files, but looks good.

File: Examples/MAX32650/ADC_MAX11261/.vscode/settings.json [hunk 1]
```diff
@@ -67,7 +67,10 @@
         "${config:MAXIM_PATH}/Libraries/MiscDrivers"
     ],
     "C_Cpp.default.defines": [
-        "${config:board}"
+        "${config:board}",
+        "TARGET=${config:target}",
+        "TARGET_REV=0x4131",
+        "__GNUC__"
```
Comment (lorne-maxim) [inline Examples/MAX32650/ADC_MAX11261/.vscode/settings.json:73]: I don't think we have any code anymore that uses 'TARGET_REV'.  We should probably start trying to phase it out.

File: Examples/MAX32655/Hello_World-riscv/Makefile [hunk 2]
```diff
@@ -44,7 +44,7 @@ all: build/$(PROJECT)-combined.elf
 build/$(PROJECT)-combined.elf: build/$(PROJECT).elf buildrv/$(PROJECT).elf arm riscv
 	@arm-none-eabi-objcopy -I elf32-littlearm build/$(PROJECT).elf -O ihex build/$(PROJECT).hex
 	@cat build/$(PROJECT).hex | sed '/^:00000001FF/d' > build/$(PROJECT)a.hex
-	@riscv-none-embed-objcopy buildrv/$(RV_PROJECT).elf -O ihex buildrv/$(PROJECT).hex
+	@riscv-none-elf-objcopy buildrv/$(RV_PROJECT).elf -O ihex buildrv/$(PROJECT).hex
 	@cat build/$(PROJECT)a.hex buildrv/$(PROJECT).hex > build/$(PROJECT)-combined.hex
 	@arm-none-eabi-objcopy -I ihex build/$(PROJECT)-combined.hex -O elf32-littlearm build/$(PROJECT)-combined.elf
```
Comment (lorne-maxim) [inline Examples/MAX32655/Hello_World-riscv/Makefile:49]: Consider using `$(ARM_PREFIX)-objcopy` and `$(RISCV_PREFIX)-objcopy`.

**Review by lorne-maxim — APPROVED** [review]

**Review by BobbyCounts — APPROVED** [review]
Changes look good, I verified the added RISCV extensions are supported in the RTL

**Pull request merged**